llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,466 @@
+  bool Modified = false;
+
+  llvm::SmallVector<const MachineBasicBlock *, 4> LoopBBs = getLoopBBs(MLI);
+
+  if (LoopBBs.empty())
+    return false;
+
+  LLVM_DEBUG(llvm::dbgs() << "*** WAW Loop Register Rewriting: "
+                          << MF.getName());
+  LLVM_DEBUG(llvm::dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopBBs) {
+    LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+               dbgs() << "\n");
+    llvm::SmallVector<MCPhysReg, 16> UsedPhyRegs;
+    llvm::SmallVector<MCPhysReg, 16> UsedVirtRegs;
+    llvm::SmallVector<MCPhysReg, 16> AlreadyReplaced;
+    llvm::SmallVector<std::array<unsigned int, 2>, 8> ReplacementRegs;
+
+    llvm::SmallVector<MCPhysReg, 16> ForbiddenPhysRegs =
+        getBBForbiddenReplacements(MBB, VRM, MRI, &TRI);
+
+    std::map<unsigned, llvm::SmallVector<const MachineInstr *, 4>>
+        UndefPartReuse;
+
+    for (const MachineInstr &MI : *MBB) {
+
+      if (isCopyElimination(MI, VRM))
+        continue;
+
+      for (const MachineOperand &Op : MI.defs()) {
+        Register Reg = Op.getReg();
+        if (!Op.getReg().isVirtual())
+          continue;
+        if (VRM.hasRequiredPhys(Op.getReg()))
+          continue;
+        if (Op.isTied())
+          continue;
+        if (!TRI.isVecOrAccRegClass(MRI.getRegClass(Reg)))
+          continue;
+        // // Subregisters can reuse Constants in the high part of the
+        // // register, therefore renaming a reused physical register can
+        // // prohibit the elimination of a move of constants, thus degrading
+        // // performance if this physical register is removed
+        if (!DisableCopySubRegHandling &&
+            isUndefReserved(Op, MI, *MBB, UndefPartReuse, VRM))
+          continue;
+
+        if (isWAWCandidate(Reg, MI, UsedPhyRegs, UsedVirtRegs, VRM, MRI)) {
+          unsigned int PreSwapPhyReg = VRM.getPhys(Reg).id();
+          bool Replaced = replaceReg(Reg, ForbiddenPhysRegs, AlreadyReplaced,
+                                     MF, VRM, LRM, MRI, LIS, TRI);
+
+          if (Replaced) {
+            MI.dump();
+            Modified = true;
+            ReplacementRegs.push_back({PreSwapPhyReg, VRM.getPhys(Reg).id()});
+          }
+
+        } else {
+          // keep track of already visted physical and virtual Regs
+          UsedPhyRegs.push_back(VRM.getPhys(Reg));
+          UsedVirtRegs.push_back(llvm::Register::virtReg2Index(Reg));
+        }
+      }
+    }
+    // dump Reassignment
+    LLVM_DEBUG(dbgs() << "WAW Replacement Overview for bb." << MBB->getNumber();
+               dbgs() << "\n");
+    for (std::array<unsigned int, 2> &RegPair : ReplacementRegs) {
+      LLVM_DEBUG(dbgs() << printReg(RegPair[0], &TRI, 0, &MRI) << " -> "
+                        << printReg(RegPair[1], &TRI, 0, &MRI) << "\n");
+    }
+    LLVM_DEBUG(dbgs() << "\n");
+  }
+  return Modified;
+}
+
+static llvm::SmallVector<const MachineBasicBlock *, 4>
```

Comment:
I think this is a good candidate for AIELoopUtils, it can be useful for other passes.

Reply:
moved it